src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,300 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of lease client
+ */
+public class EtcdLeaseImpl implements EtcdLease {
+
+    private final static int DEFAULT_TTL = 5000;
+
+    private Object lock = new Object();;
+
+    /**
+     * gRPC channel
+     */
+    private ManagedChannel channel;
+    private final LeaseGrpc.LeaseFutureStub leaseFutureStub;
+    private final LeaseGrpc.LeaseStub leaseStub;
+
+    /**
+     * Timer schedule to send keep alive request
+     */
+    private ScheduledExecutorService keepAliveSchedule;
+    private ScheduledExecutorService deadLineSchedule;
+    private Map<Long, Lease> keepAlives = new ConcurrentHashMap<>();
+
+    /**
+     * The first time interval
+     */
+    private long firstKeepAliveTimeOut = DEFAULT_TTL;
+
+    /**
+     * LeaseHandler, process exception with lease expiration
+     */
+    private EtcdLeaseHandler leaseHandler;
+
+    /**
+     * KeepAlive Request Stream, put request into this stream to keep the lease alive
+     */
+    private StreamObserver<LeaseKeepAliveRequest> keepAliveRequestStreamObserver;
+
+    /**
+     * KeepAlive Response Streamer, receive keep alive response from this stream and update the
+     * nextKeepAliveTime and deadline of the leases.
+     */
+    private StreamObserver<LeaseKeepAliveResponse> keepAliveResponseStreamObserver;
+
+    public EtcdLeaseImpl(final ManagedChannel channel) {
+
+        this(channel, null);
+    }
+
+    public EtcdLeaseImpl(final ManagedChannel channel, EtcdLeaseHandler etcdLeaseHandler) {
+
+        leaseHandler = etcdLeaseHandler;
+        /*
+        Init lease stub
+         */
+        this.channel = channel;
+        this.leaseFutureStub = LeaseGrpc.newFutureStub(this.channel);
+        this.leaseStub = LeaseGrpc.newStub(channel);
+
+        resetKeepAliveService();
+    }
+
+    /**
+     * Init the request stream to etcd
+     * start schedule to keep heartbeat to keep alive  and remove dead leases
+     */
+    public void resetKeepAliveService(){
+
+        newRequestStream();
+
+        keepAliveResponseStreamObserver = new StreamObserver<LeaseKeepAliveResponse>() {
+            @Override
+            public void onNext(LeaseKeepAliveResponse leaseKeepAliveResponse) {
+                processKeepLiveRespond(leaseKeepAliveResponse);
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                if (leaseHandler != null) {
+                    leaseHandler.onThrowable(throwable);
+                }
+            }
+
+            @Override
+            public void onCompleted() {
+
+            }
+        };
+
+        /**
+         * Start heartbeat schedule to keep alive leases
+         */
+        keepAliveSchedule = Executors.newSingleThreadScheduledExecutor();
```

Comment:
It's better to keep these two tasks as one. Both will touch the shared lease data structure and it will require additional locking. IMHO, it would be better to keep it simple